### PR TITLE
Reimplement Japanese segmenter

### LINKED ISSUE #89: Reimplement Japanese Segmenter
Reimplement Japanese segmenter using [Lindera](https://github.com/lindera-morphology/lindera).

## TODO list

- [ ] Read Contributing.md about Segmenters implementation
-  Lindera loads dictionaries at initialization
    - [ ] Ensure that Lindera is not initialized at each tokenization
    - [ ] Add a feature flag for Japanese
- [ ] use a custom config to initailize Lindera (better segmentation for search usage)
>  `TokenizerConfig { mode: Mode::Decompose(Penalty::default()), ..TokenizerConfig::default() }`
- [ ] test segmenter
> `関西国際空港限定トートバッグ すもももももももものうち` should give `["関西", "国際", "空港", "限定", "トートバッグ", " ", "すもも", "も", "もも", "も", "もも", "の", "うち"]` 

- [x] Add benchmarks

### PR DESCRIPTION
**Pull request by loiclec**
# Pull Request

## What does this PR do?
Fixes #89 

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

## Notes

* This PR adds a `japanese` feature flag which depends on `lindera` and `lindera-core`. These are the crates performing the actual segmenting. The two Japanese dictionary features of `lindera` are used. I have not checked whether both are strictly necessary.
* I have added only one unit test. The benchmarks seemed to already include Japanese text, so I assumed that there was no need to add another one.
* The existing Chinese normaliser was previously triggered on `script ==  Cj`, which meant it was also triggered on Japanese text. I have made it stricter to avoid normalising Japanese text. Now it only triggers when the language is known as `Cmn`.
* `lindera`'s `tokenize_str` method returns a `Result`. It is not clear to me on what conditions this method fails. A previous implementation of the Japanese segmenter on this repo `unwrap`ed the result.  I have done the same thing, but have low confidence on the correctness of that approach.
* Japanese compounded words are decomposed by using the following configuration:
```rust
TokenizerConfig { mode: Mode::Decompose(Penalty::default()), ..TokenizerConfig::default() }
```
* I have tried to avoid initialising `Lindera`every time by storing it in a static `Lazy`, as recommended in the dummy example.

### CODE REVIEW

File: src/segmenter/japanese.rs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+static LINDERA: Lazy<LinderaTokenizer> = Lazy::new(|| {
+    let config =
+        TokenizerConfig { mode: Mode::Decompose(Penalty::default()), ..TokenizerConfig::default() };
+    LinderaTokenizer::with_config(config).unwrap()
+});
+
+impl Segmenter for JapaneseSegmenter {
+    fn segment_str<'o>(&self, to_segment: &'o str) -> Box<dyn Iterator<Item = &'o str> + 'o> {
+        let segment_iterator = LINDERA.tokenize_str(to_segment).map(|vec| vec.into_iter()).unwrap();
+        Box::new(segment_iterator)
```
Comment (ManyTheFish) [inline src/segmenter/japanese.rs:20]:
```suggestion
        let segments = LINDERA.tokenize_str(to_segment).unwrap();
        Box::new(segments.into_iter())
```
This should work and I find it easier to read, what do you think? 😄 

About the unwrap I'll investigate later 🤔

**Review by ManyTheFish — CHANGES_REQUESTED** [review]

**Comment by ManyTheFish** [comment]
Good job @loiclec! Thank you for your contribution!
bors merge

**Comment by curquiza** [comment]
@ManyTheFish you need to approve to be able to merge 😇

**Review by ManyTheFish — APPROVED** [review]
Thank you @curquiza!
bors merge

**Pull request merged**